internals-js/src/operations.ts

```diff
@@ -1117,34 +1128,63 @@ export class NamedFragmentDefinition extends DirectiveTargetElement<NamedFragmen
+      this.expandedSelectionSetsAtTypesCache.set(type.name, cached);
     }
+    return cached;
+  }
 
-    let selectionAtType = this.expandedSelectionSetsAtTypesCache.get(type.name);
-    if (!selectionAtType) {
-      // Note that all we want is removing any top-level branches that don't apply due to the current type. There is no point
-      // in going recursive however: any simplification due to `type` stops as soon as we traverse a field. And so we don't bother. 
-      selectionAtType = expandedSelectionSet.trimUnsatisfiableBranches(type, { recursive: false });
-      this.expandedSelectionSetsAtTypesCache.set(type.name, selectionAtType);
-    }
-    return selectionAtType;
+  private computeExpandedSelectionSetAtType(type: CompositeType): SelectionSetAtType {
+    const expandedSelectionSet = this.expandedSelectionSet();
+    let selectionSet: SelectionSet;
+    // There is 2 case where we can just call `normalize` on `type` directly:
+    // 1. if `type` is an object: since `type` is a runtime of `typeCondition` then `typeCondition` is
+    //   either an interface of `type` or a union containing it.
+    // 2. if `typeCondition` is an union: that's because the only selections on a union are either `__typename`
+    //   or some other condition that intersects the union. Both are also guaranteed to be ok for `type`
+    //   since `type` intersects union.
+    if (isObjectType(type) || isUnionType(this.typeCondition)) {
+      // Note that what we want is get any simplification coming from normalizing at `type`, but any such simplication
+      // stops as soon as we traverse a field, so no point in being recursive.
+      selectionSet = expandedSelectionSet.normalize({ parentType: type, recursive: false });
+    } else {
+      // Otherwise, we just filter any top-level fragment that happens to not intersect with `type` runtimes.
+      // Note that `normalize` also do this for the case above, but `normalize` also can sometimes simplify some
+      // of the fragment that are kept, which don't apply here. And calling `normalize` would not be valid because
+      // one of its effect is to "rebase" on `type` which may be invalid here. Namely, we could have `type === I1`
+      // and the fragment be: `fragment X on I2 { f }`, and even having `runtimes(I1)` included in `runtimes(I2)`
+      // does not mean that `f` if a valid field for `I1`.
+      //
+      // Note: reminder here that this method assumes `this.canApplyAtType(type)` and that means that we already
+      // know that `typeRuntimes` is a (non-necessarily strict) subset of `conditionRuntimes`. And so if they
+      // have the same number of runtimes, then they have the same set of runtimes and no point doing any filtering.
+      const typeRuntimes = possibleRuntimeTypes(type);
+      const conditionRuntimes = possibleRuntimeTypes(this.typeCondition);
+      selectionSet = typeRuntimes.length === conditionRuntimes.length
```

Comment:
`selectionSet` in the `if` block above will be rebased on `type` due to the `SelectionSet.normalize()` call, but it looks like we'd have to manually do that rebasing here. (Not sure if it really matters, since it looks like this function is only used by `tryOptimizeSubselectionWithFragments()`.)

Reply:
You're right, and this actually made me realise that I was making this more complicated than it needed to be.

Because calling `rebaseOn(type)` on the `if` block above is actually usually incorrect. Typically, that `if` block was taken when, say, `type` was some interface `I1` while `this.typeCondition` was another interface `I2` (assuming the runtimes match what expect). But in general, you can't just rebase fields of `I2` into `I1` directly, even it happens that it "would" be alright from the POV of the underlying possible runtimes.

So really, the issue was that the `canApplyAtType` method, which was guarding calls to this `expandedSelectionSetAtType` should just not allow those cases where calling `normalize(type)` does not work. So fixed that and simplified accordingly, and I've validated that fragments are still reused as well as before.

@sachindshinde since the change is probably a bit bigger than you intended, I'll give you the chance to have a look at the last commit/check my logic before merging.

Reply:
> I'll give you the chance to have a look at the last commit/check my logic before merging.

Slightly changed my mind given circumstance: I optimistically merged, but do raise any concern if you have and we'll address.